File: scripts/analyze-exports.erl
#!/usr/bin/env escript
%% -*- erlang -*-
%%! +A0

%% Analyze Erlang export declarations
%% Finds: 1) Exported but undefined functions
%%        2) Defined but not exported functions (potential missing exports)
%%        3) Duplicate exports

-mode(compile).

main([Dir]) ->
    analyze_exports(Dir);
main(_) ->
    io:format("Usage: analyze-exports.erl <directory>~n"),
    halt(1).

analyze_exports(Dir) ->
    Files = find_erl_files(Dir),
    io:format("Found ~p Erlang files to analyze...~n~n", [length(Files)]),
    analyze_files(Files, [], []),
    io:format("~nAnalysis complete.~n"),
    halt(0).

find_erl_files(Dir) ->
    OsCmd = io_lib:format("find ~s -name '*.erl' -type f | grep -v '_build' | grep -v '_archived'", [Dir]),
    Output = os:cmd(OsCmd),
    [string:trim(F) || F <- string:split(Output, "\n", all), F =/= ""].

analyze_files([], _AllExports, _AllDefs) ->
    ok;
analyze_files([File | Rest], AllExports, AllDefs) ->
    case parse_file(File) of
        {ok, {Exports, Defs}} ->
            check_exports(File, Exports, Defs),
            analyze_files(Rest, [{File, Exports} | AllExports], [{File, Defs} | AllDefs]);
        {error, Reason} ->
            io:format(standard_error, "Error parsing ~s: ~p~n", [File, Reason]),
            analyze_files(Rest, AllExports, AllDefs)
    end.

parse_file(File) ->
    case file:read_file(File) of
        {ok, Bin} ->
            Content = binary_to_list(Bin),
            Exports = extract_exports(Content),
            Defs = extract_definitions(Content),
            {ok, {Exports, Defs}};
        {error, Reason} ->
            {error, Reason}
    end.

%% Extract exports from -export([...]) directives
extract_exports(Content) ->
    Exports = re:run(Content, "-export\\s*\\(\\s*\\[([^\\]]+)\\]\\s*\\)\\s*\\.",
                       [global, {capture, all_but_first, list}, ungreedy]),
    case Exports of
        {match, Matches} ->
            lists:flatten([parse_export_list(Match) || Match <- Matches]);
        nomatch ->
            []
    end.

%% Parse export list like "foo/1, bar/2, baz/0"
parse_export_list(String) ->
    Tokens = re:replace(String, "\\s+", "", [global, {return, list}]),
    FuncStrings = string:split(Tokens, ",", all),
    lists:filtermap(fun parse_function_arity/1, FuncStrings).

%% Parse "function_name/arity"
parse_function_arity("") ->
    false;
parse_function_arity(FuncString) ->
    case string:split(FuncString, "/", all) of
        [Name, ArityStr] ->
            case string:to_integer(ArityStr) of
                {Arity, ""} when is_integer(Arity), Arity >= 0 ->
                    {true, {list_to_atom(Name), Arity}};
                _ ->
                    false
            end;
        _ ->
            false
    end.

%% Extract function definitions
extract_definitions(Content) ->
    %% Match function definitions at module level
    %% This regex looks for "function_name(Args) ->" or "function_name(Args) when"
    %% We need to be careful to avoid matching inside comments or strings
    Lines = string:split(Content, "\n", all),
    extract_definitions_from_lines(Lines, [], in_code, 0, []).

extract_definitions_from_lines([], _Acc, _State, _Depth, Defs) ->
    lists:reverse(Defs);
extract_definitions_from_lines([Line | Rest], Acc, State, Depth, Defs) ->
    {NewState, NewDepth} = process_line_state(Line, State, Depth),
    case NewState of
        in_comment ->
            extract_definitions_from_lines(Rest, Acc, NewState, NewDepth, Defs);
        in_code ->
            case extract_function_def(Line) of
                {ok, Name, Arity} ->
                    extract_definitions_from_lines(Rest, Acc, NewState, NewDepth,
                                                   [{Name, Arity} | Defs]);
                none ->
                    extract_definitions_from_lines(Rest, Acc, NewState, NewDepth, Defs)
            end
    end.

%% Process line to determine if we're in a comment or code
process_line_state(Line, State, Depth) ->
    %% Check for attribute/record/macro definitions (no depth tracking)
    IsDirective = re:run(Line, "^-\\s*(ifdef|ifndef|else|endif|define|undef|record|macro|type|spec|opaque|callback|spec)", [caseless, {capture, none}]),
    %% Check for function/attribute start
    IsFunStart = re:run(Line, "^[a-z]\\w*\\s*\\(", [{capture, none}]),
    IsAttrStart = re:run(Line, "^-\\s*[a-z]", [{capture, none}]),
    %% Check for block start/end
    BlockStart = re:run(Line, "\\b(case|if|receive|begin|try|fun|after|catch)\\b", [caseless, {capture, none}]),
    BlockEnd = re:run(Line, "^\\s*(end|\\])\\.?", [{capture, none}]),
    HasEnd = re:run(Line, "\\bend\\b", [{capture, none}]),
    HasOpen = count_occurrences(Line, "["),
    HasClose = count_occurrences(Line, "]"),
    HasCase = re:run(Line, "\\bcase\\b", [caseless, {capture, none}]),
    HasIf = re:run(Line, "\\bif\\b", [caseless, {capture, none}]),
    HasReceive = re:run(Line, "\\breceive\\b", [caseless, {capture, none}]),
    HasTry = re:run(Line, "\\btry\\b", [caseless, {capture, none}]),
    HasBegin = re:run(Line, "\\bbegin\\b", [caseless, {capture, none}]),
    HasFun = re:run(Line, "\\bfun\\s", [caseless, {capture, none}]),

    %% Simple comment detection
    HasCommentStart = re:run(Line, "%", [{capture, none}]),
    HasBlockCommentStart = re:run(Line, "/\\*", [{capture, none}]),
    HasBlockCommentEnd = re:run(Line, "\\*/", [{capture, none}]),

    %% Calculate depth change
    StartInc = case BlockStart of
                   {match, _} -> 1;
                   nomatch -> 0
               end,
    EndInc = case HasEnd of
                 {match, _} -> 1;
                 nomatch -> 0
             end,

    NewDepth = case {HasCase, HasIf, HasReceive, HasTry, HasBegin} of
                   {{match, _}, _, _, _, _} -> Depth + 1;
                   {_, {match, _}, _, _, _} -> Depth + 1;
                   {_, _, {match, _}, _, _} -> Depth + 1;
                   {_, _, _, {match, _}, _} -> Depth + 1;
                   {_, _, _, _, {match, _}} -> Depth + 1;
                   _ -> Depth
               end,

    FinalDepth = case HasEnd of
                     {match, _} when Depth > 0 -> Depth - 1;
                     _ -> NewDepth
                 end,

    NewState = case State of
                   in_block_comment ->
                       case HasBlockCommentEnd of
                           {match, _} -> in_code;
                           nomatch -> in_block_comment
                       end;
                   in_code ->
                       case HasBlockCommentStart of
                           {match, _} -> in_block_comment;
                           nomatch -> in_code
                       end
               end,
    {NewState, FinalDepth}.

count_occurrences(_String, _Pattern) ->
    nomatch.  %% Placeholder for complex counting

%% Extract function name and arity from a line like "foo(A, B) ->"
extract_function_def(Line) ->
    Trimmed = string:trim(Line),
    case re:run(Trimmed, "^([a-z][a-zA-Z0-9_]*)\\s*\\(([^)]*)\\)\\s*(->|when)",
                 [{capture, [1, 2], list}]) of
        {match, [Name, Args]} when Name =/= "" ->
            Arity = count_arguments(Args),
            {ok, list_to_atom(Name), Arity};
        _ ->
            none
    end.

%% Count arguments in argument list
count_arguments("") ->
    0;
count_arguments(Args) ->
    %% Split by comma, but handle nested tuples, lists, etc.
    %% Simple version: count commas + 1
    Trimmed = string:trim(Args),
    case Trimmed of
        "" -> 0;
        _ -> count_commas_outside_brackets(Trimmed, 0, 0, 0, 0)
    end.

count_commas_outside_brackets([], _, _, _, Count) ->
    Count + 1;
count_commas_outside_brackets([$, | Rest], Paren, Bracket, Brace, Count) when
        Paren =:= 0, Bracket =:= 0, Brace =:= 0 ->
    count_commas_outside_brackets(Rest, Paren, Bracket, Brace, Count + 1);
count_commas_outside_brackets([$( | Rest], Paren, Bracket, Brace, Count) ->
    count_commas_outside_brackets(Rest, Paren + 1, Bracket, Brace, Count);
count_commas_outside_brackets([$) | Rest], Paren, Bracket, Brace, Count) when Paren > 0 ->
    count_commas_outside_brackets(Rest, Paren - 1, Bracket, Brace, Count);
count_commas_outside_brackets($[ | Rest], Paren, Bracket, Brace, Count) ->
    count_commas_outside_brackets(Rest, Paren, Bracket + 1, Brace, Count);
count_commas_outside_brackets($] | Rest], Paren, Bracket, Brace, Count) when Bracket > 0 ->
    count_commas_outside_brackets(Rest, Paren, Bracket - 1, Brace, Count);
count_commas_outside_brackets(${ | Rest], Paren, Bracket, Brace, Count) ->
    count_commas_outside_brackets(Rest, Paren, Bracket, Brace + 1, Count);
count_commas_outside_brackets($} | Rest], Paren, Bracket, Brace, Count) when Brace > 0 ->
    count_commas_outside_brackets(Rest, Paren, Bracket, Brace - 1, Count);
count_commas_outside_brackets([_ | Rest], Paren, Bracket, Brace, Count) ->
    count_commas_outside_brackets(Rest, Paren, Bracket, Brace, Count).

%% Check if exported functions have definitions
check_exports(File, Exports, Defs) ->
    DefSet = sets:from_list(Defs),
    ExportSet = sets:from_list(Exports),

    %% Find exported but undefined
    Undefined = sets:to_list(sets:subtract(ExportSet, DefSet)),
    %% Find defined but not exported (public interface candidates)
    NotExported = sets:to_list(sets:subtract(DefSet, ExportSet)),

    %% Check for duplicates in exports
    DuplicateExports = find_duplicates(Exports),

    %% Report issues
    case {Undefined, NotExported, DuplicateExports} of
        {[], [], []} ->
            ok;  %% No issues
        _ ->
            io:format("~s:~n", [File]),
            print_issues("  Exported but undefined", Undefined),
            print_issues("  Defined but not exported", filter_public_interface(NotExported)),
            print_issues("  Duplicate exports", DuplicateExports),
            io:format("~n")
    end.

%% Filter out internal functions that shouldn't be exported
filter_public_interface(Funcs) ->
    [F || F <- Funcs, not is_internal_function(F)].

is_internal_function({Name, _}) ->
    %% Filter out test functions, internal helpers, etc.
    NameStr = atom_to_list(Name),
    case NameStr of
        "test" ++ _ -> false;
        "_" ++ _ -> false;
        "do_" ++ _ -> false;
        "loop" -> false;
        "init" -> false;
        "handle_call" -> false;
        "handle_cast" -> false;
        "handle_info" -> false;
        "terminate" -> false;
        "code_change" -> false;
        _ -> true
    end.

find_duplicates(List) ->
    Counts = lists:foldl(fun(E, Acc) ->
                                 maps:update_with(E, fun(V) -> V + 1 end, 1, Acc)
                         end, #{}, List),
    [E || E <- maps:keys(Counts), maps:get(E, Counts) > 1].

print_issues(_, []) ->
    ok;
print_issues(Header, Issues) ->
    io:format("~s: ~p~n", [Header, Issues]).

File: scripts/analyze_exports.py
#!/usr/bin/env python3
"""
Analyze Erlang export declarations.

Finds:
1. Exported but undefined functions
2. Defined but not exported functions (potential missing exports)
3. Duplicate exports
"""

import os
import re
import sys
from collections import defaultdict
from pathlib import Path
from typing import Dict, List, Set, Tuple


def extract_exports(content: str) -> Set[Tuple[str, int]]:
    """Extract all exported functions from -export([...]) declarations."""
    exports = set()
    # Match -export([...]) where [...] contains function/arity pairs
    pattern = r'-export\s*\(\s*\[\s*([^\]]+)\s*\]\s*\)\s*\.'
    matches = re.findall(pattern, content)
    for match in matches:
        # Split by comma and parse each function/arity
        for item in re.split(r',\s*', match.strip()):
            item = item.strip()
            if not item:
                continue
            if '/' in item:
                name, arity = item.split('/', 1)
                try:
                    arity = int(arity)
                    exports.add((name, arity))
                except ValueError:
                    pass
    return exports


def extract_definitions(content: str) -> Set[Tuple[str, int]]:
    """Extract all function definitions from the module."""
    definitions = set()
    lines = content.split('\n')

    in_multiline_comment = False
    in_string = False
    in_attribute = False
    paren_depth = 0
    spec_depth = 0
    clause_continuation = False

    for line in lines:
        stripped = line.strip()

        # Handle multiline comments
        if '"""' in line or '''''' in line:
            pass  # Skip for now, simple handling
        if '/*' in line:
            in_multiline_comment = True
            comment_end = line.find('*/')
            if comment_end > line.find('/*'):
                in_multiline_comment = False
            continue
        if '*/' in line:
            in_multiline_comment = False
            continue

        # Skip comments and directives
        if in_multiline_comment or stripped.startswith('%') or not stripped:
            continue

        # Check for -export, -spec, -type, -record, etc. (skip these)
        if stripped.startswith('-'):
            # Track if we're in a multi-line spec/type
            if re.match(r'-\s*(spec|type|opaque|callback|record|macro)', stripped):
                in_attribute = True
            elif in_attribute and '.' in stripped:
                in_attribute = False
            continue

        if in_attribute:
            if '.' in stripped:
                in_attribute = False
            continue

        # Look for function definitions: Name(Args) -> or Name(Args) when
        # Must be at the start of a logical line (not indented clause)
        if not clause_continuation:
            # Match function definition pattern
            # Should start with lowercase letter, followed by optional alphanumeric,
            # then opening paren, arguments, closing paren, then -> or when
            func_match = re.match(
                r'^([a-z][a-zA-Z0-9_]*)\s*\(([^)]*)\)\s*(->|when\b)',
                stripped
            )
            if func_match:
                name = func_match.group(1)
                args_str = func_match.group(2).strip()

                # Calculate arity by counting arguments
                if not args_str:
                    arity = 0
                else:
                    arity = count_arguments(args_str)

                definitions.add((name, arity))

                # Check if this is a multi-clause function
                if ';' in stripped.split('->')[0]:
                    clause_continuation = True

        # Reset clause continuation if we see a . at end
        if stripped.endswith('.'):
            clause_continuation = False

    return definitions


def count_arguments(args_str: str) -> int:
    """Count function arguments, handling nested structures."""
    if not args_str or args_str.strip() == '':
        return 0

    args_str = args_str.strip()
    count = 1
    depth = 0
    in_string = False
    escape_next = False

    for char in args_str:
        if escape_next:
            escape_next = False
            continue

        if char == '\\':
            escape_next = True
            continue

        if char == '"' and not escape_next:
            in_string = not in_string
            continue

        if in_string:
            continue

        if char in '({[':
            depth += 1
        elif char in ')}]':
            depth -= 1
        elif char == ',' and depth == 0:
            count += 1

    return count


def find_duplicates(items: Set[Tuple[str, int]]) -> List[Tuple[str, int]]:
    """Find duplicate entries in the export list."""
    counts = defaultdict(int)
    for item in items:
        counts[item] += 1
    return [item for item, count in counts.items() if count > 1]


def is_public_function(name: str, arity: int) -> bool:
    """Check if a function should be in the public interface."""
    name_str = name if isinstance(name, str) else name

    # Skip test functions
    if name_str.startswith('test_'):
        return False
    if name_str.endswith('_test') or name_str.endswith('_tests'):
        return False
    if name_str.endswith('_suite'):
        return False

    # Skip internal functions
    if name_str.startswith('_'):
        return False
    if name_str.startswith('do_'):
        return False
    if name_str == 'loop' or name_str.endswith('_loop'):
        return False

    # Skip OTP callbacks (they should be exported via behaviours)
    otp_callbacks = {
        'init', 'handle_call', 'handle_cast', 'handle_info',
        'terminate', 'code_change', 'format_status',
        'handle_continue', 'handle_debug',
        'init', 'pre_halt', 'post_halt', 'terminate', 'handle_sync_event',
        'handle_event', 'handle_info', 'code_change', 'format_status',
        'start_link', 'start', 'stop', 'call', 'cast', 'multi_call',
        'abcast', 'bcast', 'register', 'unregister', 'whereis', 'registered',
    }
    if name_str in otp_callbacks:
        return False

    return True


def analyze_file(filepath: Path) -> Dict:
    """Analyze a single Erlang file for export issues."""
    try:
        with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
            content = f.read()
    except Exception as e:
        return {'error': str(e)}

    exports = extract_exports(content)
    definitions = extract_definitions(content)

    # Find exported but undefined
    undefined = exports - definitions

    # Find defined but not exported (potential missing exports)
    not_exported = definitions - exports
    public_not_exported = {f for f in not_exported if is_public_function(f[0], f[1])}

    # Find duplicates in exports
    # (Since we use a set, duplicates are automatically removed)
    # Need to check the raw content for duplicate export entries
    duplicate_exports = find_duplicate_exports_in_content(content)

    return {
        'exports': exports,
        'definitions': definitions,
        'undefined': undefined,
        'not_exported': public_not_exported,
        'duplicates': duplicate_exports,
    }


def find_duplicate_exports_in_content(content: str) -> Set[Tuple[str, int]]:
    """Find functions that appear multiple times in export lists."""
    all_exported = []
    pattern = r'-export\s*\(\s*\[\s*([^\]]+)\s*\]\s*\)\s*\.'
    matches = re.findall(pattern, content)
    for match in matches:
        for item in re.split(r',\s*', match.strip()):
            item = item.strip()
            if not item:
                continue
            if '/' in item:
                name, arity = item.split('/', 1)
                try:
                    arity = int(arity)
                    all_exported.append((name, arity))
                except ValueError:
                    pass

    # Count occurrences
    counts = defaultdict(int)
    for item in all_exported:
        counts[item] += 1

    return {item for item, count in counts.items() if count > 1}


def main():
    """Main entry point."""
    if len(sys.argv) < 2:
        print("Usage: analyze_exports.py <directory>...")
        sys.exit(1)

    directories = sys.argv[1:]

    issues_found = False
    total_files = 0

    for directory in directories:
        dir_path = Path(directory)
        if not dir_path.exists():
            print(f"Error: Directory not found: {directory}", file=sys.stderr)
            continue

        # Find all .erl files
        erl_files = [
            f for f in dir_path.rglob('*.erl')
            if '_build' not in str(f) and '_archived' not in str(f)
        ]

        print(f"\nAnalyzing {len(erl_files)} files in {directory}...")

        for filepath in sorted(erl_files):
            total_files += 1
            result = analyze_file(filepath)

            if 'error' in result:
                print(f"{filepath}: ERROR - {result['error']}")
                issues_found = True
                continue

            undefined = result['undefined']
            not_exported = result['not_exported']
            duplicates = result['duplicates']

            if undefined or not_exported or duplicates:
                issues_found = True
                print(f"\n{filepath}:")

                if undefined:
                    sorted_undefined = sorted(undefined, key=lambda x: (x[0], x[1]))
                    print(f"  Exported but undefined: {sorted_undefined}")

                if not_exported:
                    sorted_not_exported = sorted(not_exported, key=lambda x: (x[0], x[1]))
                    print(f"  Defined but not exported (public): {sorted_not_exported}")

                if duplicates:
                    print(f"  Duplicate exports: {sorted(duplicates)}")

    if total_files == 0:
        print("No Erlang files found.")
        sys.exit(1)

    if issues_found:
        print(f"\nIssues found in analyzed files.")
        sys.exit(1)
    else:
        print(f"\nNo issues found in {total_files} files.")
        sys.exit(0)


if __name__ == '__main__':
    main()
